In [1]:
import utils.conf as conf
import preprocessing.data_cleaner as cleaner
import preprocessing.feature_extractor as extractor
import pandas as pd
import numpy as np
%load_ext autoreload
%autoreload 2

In [2]:
train = pd.read_csv(conf.train_data_path, sep=',',usecols=['ScanCount','VisitNumber','FinelineNumber'])

In [3]:
df = train
dummies = pd.get_dummies(df.FinelineNumber)
dummy_names = ['fn_' + str(x) for x in dummies.columns]
df[dummy_names] = dummies

In [4]:
del dummies

In [6]:
del df['FinelineNumber']

SyntaxError: invalid syntax (<ipython-input-16-abb667633aec>, line 1)

In [3]:
dummies = pd.get_dummies(train.FinelineNumber)
dummies_names = [('fn_' + str(cn)) for cn in dummies.columns]

In [50]:
mask = train.FinelineNumber.value_counts() > 1000


IndexingError: Unalignable boolean Series key provided

True

In [39]:
train.FinelineNumber.nunique()

5195

In [ ]:
train.head()

In [2]:
train,_ = cleaner.read_data()

In [33]:
train, test = cleaner.clean()

In [34]:
train = extractor.aggregate_features_v2(train)
test = extractor.aggregate_features_v2(test)


KeyError: 'Column not found: TripType'

**Cross validation**

In [39]:
X = train.drop(['TripType'], axis=1)

In [40]:
from sklearn import preprocessing

tt_le = preprocessing.LabelEncoder()
tt_le.fit(train['TripType'])

LabelEncoder()

In [41]:
#from sklearn.model_selection import train_test_split
#X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0)

In [42]:
vn = train['VisitNumber'].unique()

In [43]:
from sklearn.model_selection import train_test_split
train_vn, test_vn = train_test_split(vn, test_size=0.4)
train_index = train['VisitNumber'].isin(train_vn)
y_train = tt_le.transform(train[train_index]['TripType'])
X_train = train[train_index].drop('TripType', axis=1)
y_test = tt_le.transform(train[-train_index]['TripType'])
X_test = train[-train_index].drop('TripType', axis=1)

In [49]:
X_train = extractor.aggreate_features(X_train)
X_test = extractor.aggreate_features(X_test)

** add feature **
* number of returned items per trip

In [50]:
X_train.drop('VisitNumber',axis=1, inplace=True)

In [51]:
X_test.drop('VisitNumber', axis=1, inplace=True)

In [74]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(n_estimators=100, n_jobs=4)
clf.fit(X_train,y_train)
clf_probs = clf.predict_proba(X_test)

In [75]:
import numpy as np
importances = clf.feature_importances_
std = np.std([tree.feature_importances_ for tree in clf.estimators_],
             axis=0)
indices = np.argsort(importances)[::-1]

# Print the feature ranking
print("Feature ranking:")
for f in range(X_train.shape[1]):
    print("%d. feature %d : %s (%f)" % (f + 1, indices[f], X_train.columns[indices[f]],importances[indices[f]]))
#from sklearn.model_selection import cross_val_score
#scores = cross_val_score(clf, X_test, y_test.values)
scores = clf.score(X_test,y_test)
print(scores)

from sklearn.metrics import log_loss
logloss_score = log_loss(y_test, clf_probs)
print(logloss_score)

Feature ranking:
1. feature 11 : Encoded_DepartmentDescription_has_max_by_vn (0.135551)
2. feature 13 : ScanCount_purchased_items_by_vn (0.129917)
3. feature 9 : FinelineNumber_has_max_by_vn (0.124469)
4. feature 7 : Encoded_Upc_nunique_groupby_vn (0.096143)
5. feature 6 : FinelineNumber_nunique_groupby_vn (0.080364)
6. feature 10 : Encoded_DepartmentDescription_max_by_vn (0.066343)
7. feature 5 : Encoded_DepartmentDescription_nunique_groupby_vn (0.065772)
8. feature 4 : Encoded_Weekday (0.064907)
9. feature 3 : Encoded_Upc (0.060493)
10. feature 2 : Encoded_DepartmentDescription (0.057254)
11. feature 1 : FinelineNumber (0.053116)
12. feature 8 : FinelineNumber_max_by_vn (0.032754)
13. feature 12 : ScanCount_returned_items_by_vn (0.021712)
14. feature 0 : ScanCount (0.011204)
0.608704168123
2.9213188378


Feature ranking:
1. feature 11 : Encoded_DepartmentDescription_has_max_by_vn (0.135551)
2. feature 13 : ScanCount_purchased_items_by_vn (0.129917)
3. feature 9 : FinelineNumber_has_max_by_vn (0.124469)
4. feature 7 : Encoded_Upc_nunique_groupby_vn (0.096143)
5. feature 6 : FinelineNumber_nunique_groupby_vn (0.080364)
6. feature 10 : Encoded_DepartmentDescription_max_by_vn (0.066343)
7. feature 5 : Encoded_DepartmentDescription_nunique_groupby_vn (0.065772)
8. feature 4 : Encoded_Weekday (0.064907)
9. feature 3 : Encoded_Upc (0.060493)
10. feature 2 : Encoded_DepartmentDescription (0.057254)
11. feature 1 : FinelineNumber (0.053116)
12. feature 8 : FinelineNumber_max_by_vn (0.032754)
13. feature 12 : ScanCount_returned_items_by_vn (0.021712)
14. feature 0 : ScanCount (0.011204)

0.608704168123

2.9213188378

Feature ranking:
1. feature 12 : Encoded_DepartmentDescription_has_max_by_vn (0.134543)
2. feature 5 : ScanCount_sum_groupby_vn (0.128942)
3. feature 10 : FinelineNumber_has_max_by_vn (0.123942)
4. feature 8 : Encoded_Upc_nunique_groupby_vn (0.098007)
5. feature 7 : FinelineNumber_nunique_groupby_vn (0.092697)
6. feature 6 : Encoded_DepartmentDescription_nunique_groupby_vn (0.067066)
7. feature 4 : Encoded_Weekday (0.063992)
8. feature 11 : Encoded_DepartmentDescription_max_by_vn (0.060958)
9. feature 3 : Encoded_Upc (0.060370)
10. feature 2 : Encoded_DepartmentDescription (0.057118)
11. feature 1 : FinelineNumber (0.053392)
12. feature 9 : FinelineNumber_max_by_vn (0.031842)
13. feature 13 : ScanCount_returned_items_by_vn (0.017312)
14. feature 0 : ScanCount (0.009820)

0.610022798277

2.906785231


Feature ranking:
1. feature 5 : ScanCount_sum_groupby_vn (0.146234)
2. feature 12 : Encoded_DepartmentDescription_has_max_by_vn (0.136101)
3. feature 10 : FinelineNumber_has_max_by_vn (0.126257)
4. feature 8 : Encoded_Upc_nunique_groupby_vn (0.093432)
5. feature 7 : FinelineNumber_nunique_groupby_vn (0.086363)
6. feature 6 : Encoded_DepartmentDescription_nunique_groupby_vn (0.068791)
7. feature 11 : Encoded_DepartmentDescription_max_by_vn (0.065487)
8. feature 4 : Encoded_Weekday (0.065123)
9. feature 3 : Encoded_Upc (0.060467)
10. feature 2 : Encoded_DepartmentDescription (0.055747)
11. feature 1 : FinelineNumber (0.052079)
12. feature 9 : FinelineNumber_max_by_vn (0.032862)
13. feature 0 : ScanCount (0.011058)
    
0.6123093149686295

2.64528643116

**Prepare for submission**

In [49]:
train = extractor.aggreate_features(train)
test = extractor.aggreate_features(test)

In [4]:
import models.models as models

y_train = train['TripType']
X_train = train.drop(['TripType'],axis=1)
rf_clf = models.RandomForestMultiClass()
rf_clf.fit(X_train, y_train)

RandomForest(n_estimators=50, n_jobs=4)

In [5]:
preds, classes = rf_clf.predict_proba(test)

In [ ]:
column_names = ['VisitNumber']
for cls in classes:
    column_names.append('TripType_' + str(cls))
    import pandas as pd
out = pd.DataFrame(columns=column_names)
out['VisitNumber'] = test['VisitNumber']
out.ix[:,'TripType_3':] = preds
out = out.groupby('VisitNumber').mean()
out.to_csv('submission.csv', sep=',')

**XGBoost** 

In [3]:
import xgboost as xgb

/home/wei/anaconda2/envs/pt27/lib/python2.7/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [53]:
xg_train = xgb.DMatrix(X_train, label=y_train)
xg_test = xgb.DMatrix(X_test, label=y_test)

In [54]:
param = {}
# use softmax multi-class classification
param['objective'] = 'multi:softmax'
# scale weight of positive examples
param['eta'] = 0.1
param['max_depth'] = 6
param['silent'] = 1
param['nthread'] = 4
param['num_class'] = 38

In [55]:
watchlist = [ (xg_train,'train'), (xg_test, 'test') ]
num_round = 5
bst = xgb.train(param, xg_train, num_round, watchlist )
# get prediction
pred = bst.predict( xg_test )

[0]	train-merror:0.369525	test-merror:0.384245
[1]	train-merror:0.358301	test-merror:0.373553
[2]	train-merror:0.35488	test-merror:0.371409
[3]	train-merror:0.352767	test-merror:0.369727
[4]	train-merror:0.350143	test-merror:0.368865


In [66]:
print ('predicting, classification accuracy = %f' % (1 - sum( int(pred[i]) != y_test[i] \
                                                    for i in range(len(y_test))) / float(len(y_test))))

predicting, classification accuracy = 0.631135


In [59]:
import numpy as np
# do the same thing again, but output probabilities
param['objective'] = 'multi:softprob'
bst = xgb.train(param, xg_train, num_round, watchlist );
# Note: this convention has been changed since xgboost-unity
# get prediction, this is in 1D array, need reshape to (ndata, nclass)
preds = bst.predict( xg_test ).reshape( y_test.shape[0], 38 )
classes = tt_le.inverse_transform(np.argmax(yprob, axis=1))

[0]	train-merror:0.369525	test-merror:0.384245
[1]	train-merror:0.358301	test-merror:0.373553
[2]	train-merror:0.35488	test-merror:0.371409
[3]	train-merror:0.352767	test-merror:0.369727
[4]	train-merror:0.350143	test-merror:0.368865


In [60]:
print ('predicting, classification error=%f' % (sum( int(ylabel[i]) != y_test[i] \
                                                    for i in range(len(y_test))) / float(len(y_test)) ))

predicting, classification error=0.368865


In [64]:
from sklearn.metrics import log_loss
logloss_score = log_loss(y_test, yprob)
print(logloss_score)

2.17541434135


In [50]:
import models.models as models

y_train = train['TripType']
X_train = train.drop(['TripType'],axis=1)
xgb_clf = models.XGBoostMutliClass(nthread=4, eta=0.1, max_depth=6, num_round=5, silent=1)

In [51]:
xgb_clf.fit(X_train, y_train)

XGBoostMutliClass(eta=0.1, max_depth=6, nthread=4, num_round=5, silent=1)

In [52]:
preds, classes = xgb_clf.predict_proba(test)

In [28]:
len(classes)

38

In [53]:
column_names = ['VisitNumber']
for cls in classes:
    column_names.append('TripType_' + str(cls))
    import pandas as pd
out = pd.DataFrame(columns=column_names)
out['VisitNumber'] = test['VisitNumber']
out.ix[:,'TripType_3':] = preds
out = out.groupby('VisitNumber').mean()
out.to_csv('submission.csv', sep=',')

In [40]:
out

TripType_3  TripType_4  TripType_5  TripType_6  TripType_7  \
VisitNumber                                                               
1              0.016609    0.016708    0.017749    0.016925    0.020090   
2              0.016328    0.016478    0.017546    0.016499    0.024234   
3              0.015935    0.016109    0.020005    0.016049    0.016140   
4              0.016481    0.016769    0.017753    0.016982    0.021268   
6              0.015734    0.015617    0.017004    0.015815    0.016457   
13             0.020904    0.021003    0.022240    0.021402    0.023929   
14             0.020186    0.020408    0.021789    0.020809    0.027509   
16             0.019972    0.020299    0.022070    0.020812    0.026729   
18             0.020520    0.020593    0.022166    0.021603    0.028595   
21             0.019633    0.019892    0.021501    0.019928    0.021590   
22             0.019776    0.019955    0.021676    0.019773    0.022184   
24             0.024597    0.021304    0.023197    0.069133    0.025440   
27             0.020013    0.020331    0.021486    0.020589    0.034165   
34             0.019294    0.019607    0.020722    0.019855    0.026624   
35             0.021918    0.018983    0.020670    0.170532    0.022669   
36             0.020093    0.020445    0.021569    0.020704    0.028700   
37             0.019573    0.019816    0.022383    0.019811    0.020928   
38             0.019860    0.020208    0.021393    0.020464    0.025629   
39             0.020830    0.019545    0.020976    0.094998    0.024088   
44             0.019865    0.019731    0.020361    0.020221    0.036924   
46             0.020377    0.020561    0.022557    0.020374    0.023102   
48             0.019884    0.020134    0.021687    0.020166    0.023026   
52             0.019493    0.019361    0.019980    0.019632    0.036646   
58             0.019606    0.019474    0.020988    0.019746    0.023251   
59             0.018658    0.018883    0.020400    0.018891    0.019176   
60             0.019939    0.020230    0.021607    0.020469    0.031677   
62             0.019493    0.019834    0.020925    0.020085    0.025646   
64             0.019891    0.020239    0.021426    0.020496    0.027561   
65             0.020266    0.020620    0.021754    0.020802    0.021774   
66             0.020050    0.020351    0.021614    0.020505    0.033433   
...                 ...         ...         ...         ...         ...   
191299         0.019833    0.019862    0.021561    0.020114    0.022834   
191300         0.019018    0.018979    0.019812    0.019370    0.020103   
191302         0.020175    0.020083    0.021695    0.020286    0.028833   
191303         0.020426    0.020304    0.021107    0.021017    0.029041   
191306         0.019820    0.019735    0.020567    0.019917    0.033383   
191307         0.019940    0.019932    0.020722    0.020394    0.031355   
191308         0.020381    0.020240    0.020974    0.020444    0.030580   
191309         0.021577    0.021495    0.022349    0.021733    0.022876   
191310         0.019680    0.019679    0.020566    0.019832    0.024061   
191314         0.019200    0.019021    0.019775    0.019226    0.022637   
191315         0.020089    0.020134    0.020957    0.020354    0.024961   
191316         0.019895    0.019838    0.020811    0.025908    0.030545   
191317         0.020176    0.020323    0.020967    0.020385    0.026036   
191319         0.019840    0.019789    0.020925    0.020122    0.025054   
191320         0.019864    0.019808    0.020645    0.022047    0.027796   
191321         0.019825    0.019868    0.020976    0.020069    0.023478   
191323         0.019042    0.019005    0.020129    0.019207    0.028259   
191325         0.019813    0.019748    0.020477    0.019950    0.036711   
191327         0.020006    0.019952    0.020672    0.020134    0.027068   
191328         0.062269    0.018252    0.018978    0.018390    0.023303   
191330         0.019901    0.019841   

In [8]:
import pandas as pd
df = pd.DataFrame(preds)

In [9]:
df['VisitNumber'] = test['VisitNumber']

In [11]:
df.groupby('VisitNumber').mean().to_csv('submission.csv', sep=',')

In [16]:
out.columns = column_names

ValueError: Length mismatch: Expected axis has 38 elements, new values have 653647 elements

In [22]:
len(classes)

653646